### PR TITLE
[DO NOT MERGE] Google analytics metrics

### PR DESCRIPTION
**Pull request by andrewgarner**
Adds pageview and unique pageview metrics to the metrics fact table.

**This is branched off `remove-audit-tool-code-from-cpm` . Will need to rebase and change the PR merge target once #473 is finalised.**

### CODE REVIEW

**Comment by andrewgarner** [comment]
[[3] GA: store Google Analytics metrics by day](https://trello.com/c/KwRlmarN/29-3-ga-store-google-analytics-metrics-by-day)

File: app/models/facts/metric.rb [hunk 1]
```diff
@@ -5,4 +5,6 @@ class Facts::Metric < ApplicationRecord
 
   validates :dimensions_date, presence: true
   validates :dimensions_item, presence: true
+  validates :pageviews, numericality: { only_integer: true }
```
Comment (pmanrubia) [inline app/models/facts/metric.rb]: The validation for `only_integer: true` does not allow `nil` values. 
We should be able to store metrics facts without being explicit about the GA metrics. What about adding `allow_nil: true` ?

File: app/services/google_analytics/requests/page_views_request.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ def build(args = {})
         GetReportsRequest.new.tap do |reports|
           reports.report_requests = Array.new.push(
             ReportRequest.new.tap do |request|
-              request.metrics = metrics(unique_page_views)
+              request.metrics = (metrics(pageviews) + metrics(unique_page_views))
```
Comment (pmanrubia) [inline app/services/google_analytics/requests/page_views_request.rb]: Minor: what about performing the calculation with?
```ruby
metrics(pageviews, unique_page_views)
```

File: spec/integration/publishing_spec.rb [hunk 3]
```diff
@@ -12,9 +12,9 @@
 
     date = Dimensions::Date.build(Date.today)
 
-    Facts::Metric.create(dimensions_organisation: org1, dimensions_item: item1, dimensions_date: date)
-    Facts::Metric.create(dimensions_organisation: org1, dimensions_item: item2, dimensions_date: date)
-    Facts::Metric.create(dimensions_organisation: org2, dimensions_item: item3, dimensions_date: date)
+    Facts::Metric.create(dimensions_organisation: org1, dimensions_item: item1, dimensions_date: date, pageviews: 0, unique_pageviews: 0)
```
Comment (pmanrubia) [inline spec/integration/publishing_spec.rb]: If `pageviews` and `unique_pageviews` are not mandatory, then we just can ignore them here. Using `FactoryGirl`, and defaulting metrics, could also do the trick.

File: app/models/etl/metrics.rb [hunk 4]
```diff
@@ -14,15 +14,21 @@ def process
   def create_metrics
     Facts::Metric.where(dimensions_date: date).delete_all
 
-    Dimensions::Item.where(latest: true).find_in_batches(batch_size: 50000) do |batch|
-      values = batch.pluck(:id, :organisation_id)
-      metrics = values.map do |value|
+    Dimensions::Item.select(:id, :organisation_id, :link).where(latest: true).find_in_batches(batch_size: 500) do |batch|
```
Comment (pmanrubia) [inline app/models/etl/metrics.rb]: I think we should have an independent `ETL::GAMetrics` that populates the Facts table with GA Metrics and has its own trigger (`rake` task). It should happen the same with Content Quality Metrics, as both of them should be able to run independently.

If GA is down, or having problems, or if we want to run the process for a particular day, we don't have to recreate the whole Facts Table for that day. In particular, when we are deleting the table first.

Comment (pmanrubia) [reply app/models/etl/metrics.rb]: To make it more robust, and regardless of having it split to its own process, I would create Jobs so we guarantee that a failure in one request does not prevent the whole process from working as expected.

File: app/services/google_analytics_service.rb [hunk 5]
```diff
@@ -3,7 +3,19 @@ def client
...
-  def page_views(base_paths)
+  def pageviews_for_date(date, *base_paths)
+    formatted_date = date.strftime("%Y-%m-%d")
+
+    request = GoogleAnalytics::Requests::PageViewsRequest.new.build(
+      base_paths: base_paths,
+      start_dates: [formatted_date],
+      end_date: formatted_date
+    )
+    response = client.batch_get_reports(request)
```
Comment (pmanrubia) [inline app/services/google_analytics_service.rb]: We are using a `batch_request` but we are not getting the most of it as we are only passing 1 Report Request. We are only retrieving data for 1 day, so we should be able to include until 5 reports with hundreds/thousands of paths in each request.

Comment (pmanrubia) [reply app/services/google_analytics_service.rb]: We are also impliciting batching to 500 paths per request, which is not much considering that we only request data for 1 day.

File: spec/integration/publishing_spec.rb (hunk 3 above)
Comment (andrewgarner) [reply spec/integration/publishing_spec.rb]: Fixed with eaa10d4de322c29e1eee762c867989a8558b1a07

File: app/models/facts/metric.rb (hunk 1 above)
Comment (andrewgarner) [reply app/models/facts/metric.rb]: Fixed with eaa10d4de322c29e1eee762c867989a8558b1a07

**Comment by andrewgarner** [comment]
I've extracted the schema and model changes into [their own PR](https://github.com/alphagov/content-performance-manager/pull/478) whilst I address the ETL strategy.

**Comment by andrewgarner** [comment]
Superseded by #481

**Pull request closed without merging**